lib/Core/Executor.cpp

```diff
@@ -4006,33 +4018,41 @@ void Executor::executePersistentMemoryFence(ExecutionState &state) {
   }
 }
 
-void Executor::executeCheckPersistence(ExecutionState &state,
-                                       const MemoryObject *mo) {
+/**
+ * TODO: (iangneal): we can refactor the slice constraint stuff. Currently, 
+ * there's only one set of constraints, so it's a little unnecessary.
+ */
+ExecutionState *Executor::executeCheckPersistence(ExecutionState &state,
+                                                  const MemoryObject *mo) {
   const ObjectState *os = state.addressSpace.findObject(mo);
   assert(os);
-  ObjectState *wos = state.addressSpace.getWriteable(mo, os);
-  PersistentState *ps = dyn_cast<PersistentState>(wos);
+  const PersistentState *ps = dyn_cast<PersistentState>(os);
   assert(ps);
 
-  #if 1
-  ConstraintManager orig = state.constraints;
+  ExecutionState *currState = &state;
 
   for (const ref<Expr> &sliceConstraint : ps->getConstraints()) {
+    if (!currState) break;
+
     // fprintf(stderr, "check slice for %p\n", (void*)mo->address);
     // sliceConstraint->dump();
-    ConstraintManager cm = orig;
-    cm.addConstraint(sliceConstraint);
-    state.constraints = cm;
+    currState->constraints.addConstraint(sliceConstraint);
+
+    StatePair isPersisted = fork(*currState, ps->isPersistedUnconstrained(), true);
```

Comment:
bit of a confusing name. maybe `result` or `statePair` or something. you have `notPersisted` down below